5801 enum variant doc comments are wrapped before `comment_width`

IVIURRAY opened the pull request:
Fixes https://github.com/rust-lang/rustfmt/issues/5801

ytmimi commented:
@IVIURRAY would you mind holding off on creating PRs until the bugfix is ready. You can verify that your code changes work by running the test suite locally with `cargo test`.

IVIURRAY commented:
@ytmimi yeah that is fine, will avoid raising until finished in the future.

ytmimi commented:
@IVIURRAY thanks! To clarify, I don't think there's anything wrong with opening draft PRs. However, In this particular case I don't think it's necessary if you're just adding failing test cases.

Let me know if you need any pointers for this one.

IVIURRAY commented:
Hey @ytmimi - yes, that is totally fine, I can keep the failing tests local until I'm ready to raise an MR 😄 

Also, I managed to get this MR passing with the test case from #5801 !! - Let me know if you've comments, I'm not going to say I know exactly how everything is hanging together here as this is my first MR for rustfmt. Please let me know if you want changes...

ytmimi commented:
@IVIURRAY great! I should have some time later today to review this and provide feedback

IVIURRAY commented on src/items.rs:
```diff
@@ -630,7 +630,7 @@ impl<'a> FmtVisitor<'a> {
         }
 
-        let shape = self.shape().sub_width(2)?;
+        let shape = self.shape();
```
FYI @ytmimi, this `.sub_width(2)?` has no effect on the tests that are currently in the repo passing or faililng. 

I chose to remove it as less code is better? Thought I would flag as you probably have more insight into what this might affect and I'm happy to write a test case to document it too 😄

ytmimi replied:
Please revert this change as I don't think it's related to the bug we're trying to solve for. overall the `Shape` dictates how far we're currently indented, and how much available space rustfmt has to work with in order to rewrite the current AST node on one line. There's a doc string on the `Shape` struct that you can read for more details.

We'll usually call `sub_width(n)` to reduce the width for some reason. In this case it's unclear why we're reducing the width by 2 but I wouldn't mess with this if we don't have to.

IVIURRAY replied:
Thank you so much for the brief overview, that seems to be inline with what I was thinking. I apprechiate you taking the time to explain that to me.

I've added this line back in.

ytmimi commented on src/items.rs:
```diff
@@ -656,8 +656,7 @@ impl<'a> FmtVisitor<'a> {
         let context = self.get_context();
-        // 1 = ','
-        let shape = self.shape().sub_width(1)?;
+        let shape = self.shape();
```
The `sub_width(1)` is still important for rewriting each enum variant as we need to take the trailing comma into account. The bug here is that we're apply `sub_width(1)` prematurely. We still need to reduce the width by 1 **after** we rewrite the attributes as the trailing comma doesn't impact the attributes.

Additionally, I think it would be best if these changes were [version gated](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#version-gate-formatting-changes).

probably best to change this to something like:

```rust
let attrs_str = if context.config.version() == Version::Two {
    field.attrs.rewrite(&context, shape)?
} else {
   // Version::One formatting that was off by 1. See issue #5801
   field.attrs.rewrite(&context, shape.sub_width(1)?)?
};

// sub_width(1) to take the trailing comma into account
let shape = shape.sub_width(1)?;
```

ytmimi replied:
wait, this is only an issue when `wrap_comments=true` is set, right? In that case we don't need the version gate since `wrap_comments=true` is an unstable option. I'd then expect the code change to look something like:

```rust
let attrs_str = field.attrs.rewrite(&context, shape)?;
// sub_width(1) to take the trailing comma into account
let shape = shape.sub_width(1)?;
```

ytmimi replied:
As I was thinking about this more I'm now 100% convinced this change needs to be version gated as attribute formatting is also influenced by stable configuration options like `attr_fn_like_width` and `use_small_heuristics`.

Consider this extended example from what you have in the test case.

```rust
pub enum Severity {
    /// But here, this comment is 120 columns wide and the formatter wants to split it up onto two separate lines still.
    Error,
    /// This comment is 119 columns wide and works perfectly. Lorem ipsum. lorem ipsum. lorem ipsum. lorem ipsum lorem.
    Warning,
    #[something(AAAAAAAAAAAAA, BBBBBBBBBBBBBB, CCCCCCCCCCCCCCCC, DDDDDDDDDDDDD, EEEEEEEEEEEE, FFFFFFFFFFF, GGGGGGGGGGG)]
    AttrsWillWrap,
    #[something_else(hhhhhhhhhhhhhhhh, iiiiiiiiiiiiiiii, jjjjjjjjjjjjjjj, kkkkkkkkkkkkk, llllllllllll, mmmmmmmmmmmmmm)]
    AttsWontWrap,
}
```

Using a recent nightly `rustfmt 1.7.0-nightly (89e2160c 2023-12-27)` with the input above and these configuration options:

```toml
comment_width = 120
max_width = 120
wrap_comments = true
version = "One"
attr_fn_like_width = 120
```

produces this formatting:

```rust
pub enum Severity {
    /// But here, this comment is 120 columns wide and the formatter wants to split it up onto two separate lines
    /// still.
    Error,
    /// This comment is 119 columns wide and works perfectly. Lorem ipsum. lorem ipsum. lorem ipsum. lorem ipsum lorem.
    Warning,
    #[something(
        AAAAAAAAAAAAA,
        BBBBBBBBBBBBBB,
        CCCCCCCCCCCCCCCC,
        DDDDDDDDDDDDD,
        EEEEEEEEEEEE,
        FFFFFFFFFFF,
        GGGGGGGGGGG
    )]
    AttrsWillWrap,
    #[something_else(hhhhhhhhhhhhhhhh, iiiiiiiiiiiiiiii, jjjjjjjjjjjjjjj, kkkkkkkkkkkkk, llllllllllll, mmmmmmmmmmmmmm)]
    AttsWontWrap,
}
```

IVIURRAY replied:
I've updated the code to include the version gate. Would you also like me to update the test cases to match what you've as an example here?

ytmimi commented on tests/source/configs/comment_width/5801.rs:
```diff
@@ -0,0 +1,16 @@
+// rustfmt-comment_width: 120
```
If there's not change between the input and the output then you can simple remove the `source` file.

IVIURRAY replied:
There is no change, I will drop the `/source` file

ytmimi commented on tests/source/configs/comment_width/5801.rs:
```diff
@@ -0,0 +1,16 @@
+// rustfmt-comment_width: 120
+// rustfmt-wrap_comments: true
+// rustfmt-unstable_features: true
```
You should be able to omit the `ustable_features` option when writing tests.

IVIURRAY replied:
Ah ok 👍 I just lifted the config from the issue, let me remove that now

ytmimi commented:
@IVIURRAY great work on this so far. I hope you're getting a better sense of what it's like to work on rustfmt! I'd like you to add two additional test cases, and move the test cases around.

* a test case for `version=One` which shows that we prematurely wrap. It's always good to have a test for the stable formatting (even if it's buggy)
* a separate test case for the wrapping attributes. I'd prefer if this was a distinct test file to highlight that we can trigger the issue with just stable options.
* Also, can you move these test cases into a `tests/{source | target}/issue_5801` folder. Feel free to name these test cases whatever you think is best. Here's some inspiration on the names:
    * `comment_unexpectedly_wraps_before_max_width` (`Version::One` test case for doc comments)
    * `comment_does_not_wrap_within_max_width` (`Version::Two` test case for doc comments)
    * `attribute_unexpectedly_wraps_before_max_width` (`Version::One` test case for attributes)
    * `attribute_does_not_wrap_within_max_width` (`Version::Two` test case for attributes)

ytmimi commented:
Also, probably best to squash all of these commits down into a single commit. Feel free to do that on your end, otherwise I'll handle that when merging.

IVIURRAY commented:
@ytmimi ok, I'm a little confused here. My test for [/attribute_does_not_wrap_within_max_width.rs](https://github.com/rust-lang/rustfmt/blob/fb36c1438c5f451fb4c4b2ae90c4fab41bf9f345/tests/target/issue-5801/attribute_does_not_wrap_within_max_width.rs) is unstable and I don't know why.

Running this test in my IDE passes and then fails and then passes. You need to just keep running it a few times. When I pushed from my IDE it passed then running on github actions failed. This caused me to notice the same was happening locally. I'm verey confused as to why this would be. Do you've any insight?

```rust
Mismatch at tests/source/issue-5801/attribute_does_not_wrap_within_max_width.rs:3:
 // rustfmt-attr_fn_like_width: 120
 
 pub enum Severity {
-    #[something(AAAAAAAAAAAAA, BBBBBBBBBBBBBB, CCCCCCCCCCCCCCCC, DDDDDDDDDDDDD, EEEEEEEEEEEE, FFFFFFFFFFF, GGGGGGGGGGG)]
+    #[something(
+        AAAAAAAAAAAAA,
+        BBBBBBBBBBBBBB,
+        CCCCCCCCCCCCCCCC,
+        DDDDDDDDDDDDD,
+        EEEEEEEEEEEE,
+        FFFFFFFFFFF,
+        GGGGGGGGGGG
+    )]
     AttrsWillWrap,
     #[something_else(hhhhhhhhhhhhhhhh, iiiiiiiiiiiiiiii, jjjjjjjjjjjjjjj, kkkkkkkkkkkkk, llllllllllll, mmmmmmmmmmmmmm)]
     AttsWontWrap,
```

IVIURRAY commented:
Also, just to let you know I squashed the commits, rebased master and created an `issue-5801` folder (I know you said `issue_5801`but the other folders followed that format in `/target`.

ytmimi commented on tests/target/issue-5801/comment_unexpectedly_wraps_before_max_width.rs:
I think you're missing the `source` file for the version One test case.

IVIURRAY replied:
Ok - I'll update 👍

ytmimi commented:
> Running this test in my IDE passes and then fails and then passes. You need to just keep running it a few times. When I pushed from my IDE it passed then running on github actions failed. This caused me to notice the same was happening locally. I'm very confused as to why this would be. Do you've any insight?

I'm just as confused. That shouldn't happen. I wrote a little script to run cargo test over and over again:

```bash
#!/bin/bash
set -e

for i in {1..20};
do
    cargo test -q &>/dev/null
    if [ $? -eq 0 ]; then
        echo "PASS $i"
    else
        echo "FAIL $i"
    fi
done
```

I'm unable to reproduce the failure locally when running `cargo test`. How are you running the tests? Maybe there are certain steps that are needed to reproduce the issue.

Also, I find it odd that this is failing in CI for the `windows / (x86_64-pc-windows-msvc, nightly)` and `windows / (x86_64-pc-windows-gnu, nightly)` targets.

ytmimi commented on tests/target/issue-5801/attribute_unexpectedly_wraps_before_max_width.rs:
```diff
@@ -0,0 +1,18 @@
+    )]
+    AttrsWillWrap,
+    #[something_else(hhhhhhhhhhhhhhhh, iiiiiiiiiiiiiiii, jjjjjjjjjjjjjjj, kkkkkkkkkkkkk, llllllllllll, mmmmmmmmmmmmmm)]
+    AttsWontWrap,
```
Let's fix the typo. ` AttsWontWrap` should be `AttrsWontWrap`

IVIURRAY replied:
Updated!

IVIURRAY commented:
> > Running this test in my IDE passes and then fails and then passes. You need to just keep running it a few times. When I pushed from my IDE it passed then running on github actions failed. This caused me to notice the same was happening locally. I'm very confused as to why this would be. Do you've any insight?
> 
> I'm just as confused. That shouldn't happen. I wrote a little script to run cargo test over and over again:
> 
> ```shell
> #!/bin/bash
> set -e
> 
> for i in {1..20};
> do
>     cargo test -q &>/dev/null
>     if [ $? -eq 0 ]; then
>         echo "PASS $i"
>     else
>         echo "FAIL $i"
>     fi
> done
> ```
> 
> I'm unable to reproduce the failure locally when running `cargo test`. How are you running the tests? Maybe there are certain steps that are needed to reproduce the issue.
> 
> Also, I find it odd that this is failing in CI for the `windows / (x86_64-pc-windows-msvc, nightly)` and `windows / (x86_64-pc-windows-gnu, nightly)` targets.

I've recorded a video of me just running `cargo test` from the cli and getting different results  - see [here](https://youtu.be/BLa5ddts5jU)

Honestly, I think if we re-run the pipeline it won't be consistent. I think I just got lucky there with some build systems passing.

ytmimi commented:
@IVIURRAY thanks for the video. I was able to reproduce the error when running the `idempotence_tests`. My hunch is that the issue is related to setting the configs inline with the rustfmt comments like `//rustfmt-{key}: {value}` in [`read_config`](https://github.com/rust-lang/rustfmt/blob/75e3172aaaff1ff6be42f36777a6814c0a72320f/src/test/mod.rs#L691-L712). From what I can tell `read_significant_comments` returns a `HashMap<String, String>`, and when we try to iterate over it we don't necessarily iterate over the configurations in the same order that they're listed in the file. Sometimes we set the `max_width=120` before we set `attr_fn_like_width=120`, and in that case everything works out. In other cases we try to set `attr_fn_like_width=120` first, and that fails since we don't allow any width configuration to exceed the `max_width`, which is still set to its default of 100. so `attr_fn_like_width` would also get set to 100.

Just to be totally clear, this isn't an issue I want to address in this PR, and I think I need to do some more investigating on my end.

To explain further, In most cases setting the configuration values inline with the rustfmt comments isn't an issue because the configuration values don't depend on each other, but in this case all the width configurations depend on the `max_width`.

I think the workaround here is to use the `// rustfmt-config:` comment, where we can specify the name of a configuration file listed in the `tests/config` directory. When we read the config from a `toml` file we don't have this same issue.

I think we can create two new files, let's call them `tests/config/issue-5801-v1.toml` and `tests/config/issue-5801-v2.toml`.

Then in each test file you can replace the inline configs with `// rustfmt-config: issue-5801-v1.toml` or`// rustfmt-config: issue-5801-v2.toml`

IVIURRAY commented:
Ah, you genius @ytmimi! - That makes total sense to me, I can totally understand how that is is causing the issue. I thought I had broken something even worse!

I've moved the config into a .toml file now and checks are passing 🎉 

Would you like me to write this up as an issue?...maybe this could be the next thing I can work on. You might've something else that would be easier for me though, wait to hear.

ytmimi reviewed (approved):
Thank you for the kind words @IVIURRAY! I thought this one would be simple, but I couldn't have anticipated all the issues that would have popped up. I'm really glad that you continued to work through them all.

Thank you for your first contribution to rustfmt 🎉

It would be great if you could create a new issue to document the testing problem we ran into. I don't think you need to continue looking into it right now. It's a bit of an edge case that I don't think we'll run into soon. I've got a rough idea on how to correct this problem, but as I mentioned, there are a few things I want to double check first.

I don't have another issue for you off the top of my head, but I'm happy to look through the backlog to see if there's another one that could ease you into the codebase.

IVIURRAY commented:
Wahoo! - Thanks for the approval and merge @ytmimi, glad I could contribute! I learnt a lot even if most of that was understanding the Rust toolchain 😆

I've created an issue to capture unstable test https://github.com/rust-lang/rustfmt/issues/6011

I would like another issue to investigate if you've on that you think would be suitable for me 😄 I'll wait to hear from you.

ytmimi commented:
As long as you learned something then it was all worth it! I appreciate you following up and creating the issue. I'll checkout the backlog over the next day or so and tag you on an issue that I think would be another good one for you to take on!
